common/common.c

```diff
@@ -94,6 +94,30 @@ char *mp_format_time(double time, bool fractions)
     return mp_format_time_fmt(fractions ? "%H:%M:%S.%T" : "%H:%M:%S", time);
 }
 
+char *mp_format_double(void *talloc_ctx, double val, int precision,
+                       bool plus_sign, bool percent_sign, bool trim)
+{
+    bstr str = {0};
+    const char *fmt = plus_sign ? "%+.*f" : "%.*f";
+    bstr_xappend_asprintf(NULL, &str, fmt, precision, val);
+    size_t pos = str.len;
+    if (trim) {
+        while (--pos && str.start[pos] == '0')
+            str.len--;
+        if (str.start[pos] == '.')
+            str.len--;
+    }
+    if (percent_sign) {
+        if (pos + 1 >= str.len) {
+            bstr_xappend(NULL, &str, bstr0("%"));
+        } else {
+            str.start[pos + 1] = '%';
```

Comment:
`pos` can be `== str.len` here

Reply:
> pos == `str.len` here,

Not possible, because `pos + 1 >= str.len` check above. But I removed this check for clarity, `bstr_xappend` will not realloc if not needed anyway, so this micro optimization is not worth readability lose.

Reply:
Generate some tests to solidify this function behavior.

Reply:
Yes you're right, my bad.